core/src/main/java/io/questdb/mp/SimpleWaitingLock.java

```diff
@@ -0,0 +1,89 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.mp;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
+
+public class SimpleWaitingLock {
+    private final AtomicBoolean lock = new AtomicBoolean(false);
+    private volatile Thread waiter = null;
+
+    public boolean isLocked() {
+        return lock.get();
+    }
+
+    public void lock() {
+        this.waiter = Thread.currentThread();
+        while (true) {
+            while (lock.get()) {
+                LockSupport.park();
+            }
+            if (!lock.getAndSet(true)) {
+                return;
+            }
+        }
+    }
+
+    public boolean tryLock(long timeout, TimeUnit unit) {
+        if (tryLock()) {
+            return true;
+        }
+
+        if (timeout <= 0L) {
+            return false;
+        }
+
+        long nanos = unit.toNanos(timeout);
+        this.waiter = Thread.currentThread();
+        while (true) {
+            long start = System.nanoTime();
+            LockSupport.parkNanos(nanos);
+            long elapsed = System.nanoTime() - start;
+
+            if (elapsed < nanos) {
+                if (tryLock()) {
+                    return true;
+                }
+            } else {
+                return false;
+            }
+        }
+    }
+
+    public boolean tryLock() {
+        return lock.compareAndSet(false, true);
+    }
+
+    public void unlock() {
```

Comment:
Should unlock actually be idempotent?
If one calls `unlock` in a non-locked state I'd say it's a bug as it points to bad API use - Especially since the lock is non-reentrant.
I don't see checks for this here.

Reply:
Fixed in https://github.com/questdb/questdb/pull/2753/commits/5f156c3c3cb96e3f3cd476ab38e07d10b94b2686
I'm reporting the wrong state via the return value rather than throwing an exception to avoid all those nested handlers.

Reply:
I'm not sure that's particularly useful. What is one supposed to do with a `false` return?
In practice I don't think anyone's going to do anything with it in practice.

I think raising an unchecked exception (`IllegalStateException` maybe?) is a better API choice.

If this were C I'd say call `abort()` :-)

Reply:
Fixed in https://github.com/questdb/questdb/pull/2753/commits/2c0aa9102357a304909b50cd09fd6f4b82de82a7